flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -0,0 +1,296 @@
+import org.apache.flink.runtime.checkpoint.JobManagerTaskRestore;
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.deployment.TaskDeploymentDescriptor.MaybeOffloaded;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.Execution;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.ExecutionEdge;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.executiongraph.ExecutionGraphException;
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.executiongraph.IntermediateResult;
+import org.apache.flink.runtime.executiongraph.IntermediateResultPartition;
+import org.apache.flink.runtime.executiongraph.JobInformation;
+import org.apache.flink.runtime.executiongraph.TaskInformation;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.IntermediateDataSetID;
+import org.apache.flink.runtime.jobmaster.LogicalSlot;
+import org.apache.flink.runtime.shuffle.ShuffleDescriptor;
+import org.apache.flink.runtime.shuffle.UnknownShuffleDescriptor;
+import org.apache.flink.types.Either;
+import org.apache.flink.util.SerializedValue;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Factory of {@link TaskDeploymentDescriptor} to deploy {@link org.apache.flink.runtime.taskmanager.Task} from {@link Execution}.
+ */
+public class TaskDeploymentDescriptorFactory {
+	private final ExecutionAttemptID executionId;
+	private final int attemptNumber;
+	private final MaybeOffloaded<JobInformation> serializedJobInformation;
+	private final MaybeOffloaded<TaskInformation> taskInfo;
+	private final JobID jobID;
+	private final boolean lazyScheduling;
+	private final int subtaskIndex;
+	private final ExecutionEdge[][] inputEdges;
+
+	private TaskDeploymentDescriptorFactory(
+			ExecutionAttemptID executionId,
+			int attemptNumber,
+			MaybeOffloaded<JobInformation> serializedJobInformation,
+			MaybeOffloaded<TaskInformation> taskInfo,
+			JobID jobID,
+			boolean lazyScheduling,
+			int subtaskIndex,
+			ExecutionEdge[][] inputEdges) {
+		this.executionId = executionId;
+		this.attemptNumber = attemptNumber;
+		this.serializedJobInformation = serializedJobInformation;
+		this.taskInfo = taskInfo;
+		this.jobID = jobID;
+		this.lazyScheduling = lazyScheduling;
+		this.subtaskIndex = subtaskIndex;
+		this.inputEdges = inputEdges;
+	}
+
+	public TaskDeploymentDescriptor createDeploymentDescriptor(
```

Comment:
This method could be replaced by the below `createDeploymentDescriptor` with specific parameters, I think it is no need to maintain multiple forms for creating TDD.